contrib/mesos/pkg/queue/delay/indexed_delay_queue.go

```diff
@@ -0,0 +1,248 @@
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/pivotal-golang/clock"
+
+	"k8s.io/kubernetes/contrib/mesos/pkg/queue"
+	"k8s.io/kubernetes/contrib/mesos/pkg/queue/priority"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+// IndexedDelayQueue is a thread-safe, time-based queue with uniquely identified items.
+//
+// Adding or offering an item with an ID matching an item still in the queue will
+// invoke the configured SchedulingPolicy and supplied ValueReplacementPolicy to
+// determine the desired behavior.
+//
+// Items can only be popped after their event time has been reached.
+// Use Add to specify a delay duration and have the event time calculated.
+// Use Offer to specify a specific event time.
+type IndexedDelayQueue struct {
+	*Queue
+	// items mapped by UniqueID - item set must match queue contents
+	items            map[string]priority.Item
+	schedulingPolicy SchedulingPolicy
+	clock            clock.Clock
+}
+
+func NewIndexedDelayQueue(clock clock.Clock) *IndexedDelayQueue {
+	return &IndexedDelayQueue{
+		Queue: NewDelayQueue(clock),
+		items: map[string]priority.Item{},
+		clock: clock,
+	}
+}
+
+// Add inserts an uniquely identified item into the queue,
+// provided another item with the same ID is not already present.
+func (q *IndexedDelayQueue) Add(d UniqueDelayed, rp ValueReplacementPolicy) {
+	q.push(&indexedAddedItem{
+		Item: priority.NewItem(d, NewDelayedPriority(d, q.clock)),
+		id:   d.GetUID(),
+	}, rp)
+}
+
+// indexedAddedItem wraps an item that was added to the queue.
+type indexedAddedItem struct {
+	priority.Item
+	id string
+}
+
+func (di indexedAddedItem) GetUID() string {
+	return di.id
+}
+
+// Push puts the item back into the queue (with the same ID)
+func (di *indexedAddedItem) Push(queue heap.Interface) {
+	dq := queue.(*IndexedDelayQueue)
+	dq.push(di, KeepExisting)
+}
+
+func (q *IndexedDelayQueue) Offer(d UniqueScheduled, rp ValueReplacementPolicy) bool {
+	if p, ok := NewScheduledPriority(d); ok {
+		q.push(&indexedOfferedItem{
+			Item: priority.NewItem(d, p),
+			id:   d.GetUID(),
+		}, rp)
+		return true
+	}
+	return false
+}
+
+func (q *IndexedDelayQueue) push(newItem UniqueItem, rp ValueReplacementPolicy) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	id := newItem.GetUID()
+	item, exists := q.items[id]
+	if !exists {
+		item = newItem
+		heap.Push(q.Queue, item)
+		q.items[id] = item
+	} else {
+		// replace existing item
+		newValue := rp.Value(item.Value(), newItem.Value())
+		oldPriority := item.Priority().(Priority)
+		newPriority := newItem.Priority().(Priority)
+		newPriority = q.schedulingPolicy.EventTime(oldPriority, newPriority)
+		item = priority.NewItem(newValue, newPriority)
+
+		switch i := newItem.(type) {
+		case *indexedAddedItem:
+			item = &indexedAddedItem{Item: item, id: i.id}
+		case *indexedOfferedItem:
+			item = &indexedOfferedItem{Item: item, id: i.id}
+		default:
+			panic(fmt.Sprintf("unsupported newItem type: %v", reflect.TypeOf(i)))
+		}
+
+		heap.Fix(q.Queue, item.Index())
```

Comment:
ahh, right. I was mixing up the items map with the queue.
I wanted the items in the queue to be immutable, because it makes the interface and assumptions easier. This is the only mutation. So I guess there are three options:
1. allow item mutability
2. remove the old item and add a new one
3. rewrite the item pointer used by the queue to point to the new item